navigator:
  ros__parameters:
    # Input data directory for ModuleVO
    input_data_dir: "~/ASRL/dataset/2016_in_the_dark"
    # Directory to save results to
    results_dir: "~/ASRL/temp/vtr3/offline/2016_in_the_dark/results"
    # Data directory for chunked data (contains run_000***)
    data_dir: "~/ASRL/temp/vtr3/offline/2016_in_the_dark/output"

    # Run directory for the simulated camera (for playback)
    sim_run: "run_000001"
    # The full name of the input camera data stream
    stream_name: /front_xb3/processed_stereo_images
    # The sensor frame from the urdf (also defines the rig name)
    sensor_frame: front_xb3
    # The control frame from the urdf
    control_frame: base_link

    start_index: 1
    stop_index: 3000

    ############ Tactic configuration ############
    tactic:
      type: basic # 'basic' or 'parallel'
      num_threads: 4 # only applies to the parallel tactic
      keyframe_parallelization: false
      keyframe_skippable: false
      extrapolate_timeout: 1.0
      use_existing_graph: false

    # # Misc configuration
    # scale: 1.0
    # throttle: false

    ############ converter module settings ############
    converter:
      # type: image_conversion
      conversion:
        # conversions: []
        # conversions: ["RGB_TO_GRAYSCALE"]
        conversions: ["RGB_TO_GRAYSCALE", "RGB_TO_COLOR_CONSTANT"]

      # extraction module
      extraction:
        type: conversion_extraction

        # Conversion-Extraction module
        # conversions: []
        # conversions: ["RGB_TO_GRAYSCALE"]
        conversions: ["RGB_TO_GRAYSCALE", "RGB_TO_COLOR_CONSTANT"]
        color_constant:
          weights: [0.43]

        extractor:
          type: ASRL_GPU_SURF # ASRL_GPU_SURF, OPENCV_ORB
          # channels: ["grayscale"]
          channels: ["grayscale", "cc_0.430000"]

          # converter/extraction/extractor/patchSize: 64
          # ORB extraction module
          orb:
            num_detector_features: 10000
            num_binned_features: 2000
            fastThreshold: 5
            x_bins: 8
            y_bins: 6
            edgeThreshold: 16
            matcher:
              stereo_descriptor_match_thresh: 0.3
              stereo_y_tolerance: 2.0
              scale_x_tolerance_by_y: false

          # SURF extraction module
          surf:
            upright_flag: true
            threshold: .000001
            nOctaves: 4
            nIntervals: 4
            initialScale: 1.5
            edgeScale: 1.5
            l1: 3.f/1.5f
            l2: 5.f/1.5f
            l3: 3.f/1.5f
            l4: 1.f/1.5f
            initialStep: 1
            targetFeatures: 1000
            detector_threads_x: 16
            detector_threads_y: 16
            regions_horizontal: 16
            regions_vertical: 16
            regions_target: 1000
            stereoDisparityMinimum: 0.0
            stereoDisparityMaximum: 64
            stereoCorrelationThreshold: .79
            stereoYTolerance: 0.9
            stereoScaleTolerance: .9

      # triangulation module
      triangulation:
        visualize_features: false
        visualize_stereo_features: false

    # ############ Quick VO settings ############

    # # matcher module
    # quick_vo/matcher/check_response: true
    # quick_vo/matcher/tight_matching_pixel_thresh: 20
    # quick_vo/matcher/matching_pixel_thresh: 400
    # quick_vo/matcher/tight_matching_x_sigma: 0.6
    # quick_vo/matcher/tight_matching_y_sigma: 0.6
    # quick_vo/matcher/tight_matching_theta_sigma: 0.6
    # quick_vo/matcher/max_point_depth: 500
    # quick_vo/matcher/descriptor_thresh: 0.1
    # quick_vo/matcher/parallel_threads: 8
    # quick_vo/matcher/prediction_method: se3
    # quick_vo/matcher/visualize_feature_matches: false

    # # RANSAC module
    # quick_vo/ransac/iterations: 1500
    # quick_vo/ransac/threshold: 8.0
    # quick_vo/ransac/early_stop_min_inliers: 2000
    # quick_vo/ransac/min_inliers: 50
    # quick_vo/ransac/visualize_ransac_inliers: true

    # # Keyframe optimisation
    # quick_vo/steam/iterations: 5
    # quick_vo/steam/pose_prior_enable: false
    # quick_vo/steam/ang_vel_std_dev_z: 0.2
    # quick_vo/steam/verbose: false
    # quick_vo/steam/use_T_q_m_prior: false
    # #quick_vo/steam/solver_type: "LevenburgMarquardt"
    # quick_vo/steam/depth_prior_enable: false
    # quick_vo/steam/trajectory_smoothing: true
    # quick_vo/steam/lin_acc_std_dev_x: 10.0
    # quick_vo/steam/lin_acc_std_dev_y: 10.0
    # quick_vo/steam/lin_acc_std_dev_z: 10.0
    # quick_vo/steam/ang_acc_std_dev_x: 1.0
    # quick_vo/steam/ang_acc_std_dev_y: 1.0
    # quick_vo/steam/ang_acc_std_dev_z: 1.0
    # quick_vo/steam/velocity_prior: false
    # quick_vo/steam/lin_vel_mean_x: 4.0
    # quick_vo/steam/lin_vel_mean_y: 0.0
    # quick_vo/steam/lin_vel_mean_z: 0.0
    # quick_vo/steam/ang_vel_mean_x: 0.0
    # quick_vo/steam/ang_vel_mean_y: 0.0
    # quick_vo/steam/ang_vel_mean_z: 0.0
    # quick_vo/steam/lin_vel_std_dev_x: 8.0
    # quick_vo/steam/lin_vel_std_dev_y: 3.0
    # quick_vo/steam/lin_vel_std_dev_z: 0.5
    # quick_vo/steam/ang_vel_std_dev_x: 0.5
    # quick_vo/steam/ang_vel_std_dev_y: 0.5
    # quick_vo/steam/perform_planarity_check: false
    # quick_vo/steam/plane_distance: 20.0
    # quick_vo/steam/min_point_depth: 0.0
    # quick_vo/steam/max_point_depth: 200.0
    # # Vertex test module
    # quick_vo/vertex_test/match_threshold_min_count: 150
    # quick_vo/vertex_test/match_threshold_fail_count: 80

    # ############ Refined VO settings ############

    # # Steam module
    # refined_vo/steam/trajectory_smoothing: true
    # refined_vo/steam/velocity_prior: false
    # refined_vo/steam/verbose: false
    # refined_vo/steam/use_T_q_m_prior: false
    # #refined_vo/steam/solver_type: "LevenburgMarquardt"
    # refined_vo/steam/depth_prior_enable: true
    # refined_vo/steam/depth_prior_weight: 1000000000
    # refined_vo/steam/max_point_depth: 800.0
    # refined_vo/steam/min_point_depth: 1.0
    # refined_vo/steam/perform_planarity_check: false
    # refined_vo/steam/plane_distance: 20.0
